domains/client/domain-operator/src/domain_bundle_proposer.rs

```diff
@@ -11,23 +11,63 @@ use sp_domains::{
     BundleHeader, DomainId, DomainsApi, ExecutionReceipt, HeaderHashingFor, ProofOfElection,
 };
 use sp_runtime::traits::{Block as BlockT, Hash as HashT, Header as HeaderT, NumberFor, One, Zero};
+use sp_runtime::Percent;
 use sp_transaction_pool::runtime_api::TaggedTransactionQueue;
 use sp_weights::Weight;
+use std::collections::HashSet;
 use std::marker::PhantomData;
 use std::sync::Arc;
 use std::time;
 use subspace_core_primitives::U256;
 use subspace_runtime_primitives::Balance;
 
-pub struct DomainBundleProposer<Block, Client, CBlock, CClient, TransactionPool> {
+/// If the bundle utilization is below `BUNDLE_UTILIZATION_THRESHOLD` we will attempt to push
+/// at most `MAX_SKIPPED_TRANSACTIONS` number of transactions before quitting for real.
+const MAX_SKIPPED_TRANSACTIONS: usize = 8;
```

Comment:
Instead of retrying till `MAX_SKIPPED_TRANSACTIONS` time to try to include a tx in bundle, why not sort ready extrinsics by their weight. AFAIK, this would be light weight runtime call and then when we reach the extrinsic for which we cannot include, we can just stop iterating.

Reply:
There are multiple things to consider here. TXpool would just give ready transactions AFAIK which means if a user has multiple transactions with incremental nonces, we might not be able to get them in the first pass. Would need to check the code path for this.

I think sorting all the ready_transactions by weight makes sense and if we still have some space, then picking the next set of transactions which will be moved to ready if the first set were imported would be helpful too

Reply:
The bundle can be full due to both weight and size, sorting transactions by weight doesn't solve all the problems. And fundamentally, the transaction should only sorted by the tip provided by the user.

>then picking the next set of transactions which will be moved to ready if the first set were imported would be helpful too

From my understanding, tx only move from the future queue to the ready queue after a new block is imported.

Reply:
@NingLin-P what we can do is sort it by (priority, tip, weight,  size) key.  This can eliminate both need to retry for 8 times and possibility of users prevening heavy tx to be included by spanming light weight tx. (if we only sort by weight).

Reply:
The tx pool already sorts the tx by priority and tip, from a UX standpoint the next thing that should be used to sort is the order of arrival of each tx.

The purpose of this change is to maximize bundle utilization thus we only try more tx if the current bundle utilization is less than 95%, and IMO trying for 8 more tx is not really a heavy operation, in the worst situation where none of the tx can fit into the bundle, the cost just 8 runtime API call for query the tx weight, and IMHO it doesn't worth the effort to support sort by weight (without considering the UX impact) which may require invasive change to the upstream.

Reply:
Hmm. I thought it would be easier to ask Substrate tx pool to custom sort the tx. Just looked at the code and you are right. 

In that case, your approach does make sense, even though it is less elegant.